// @halva-suite/halva-cli/build/main/TestRunner/Config/HalvaTestConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HalvaTestConfig = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const FindConfigFile_1 = require("../../Configure/FindConfigFile");
class HalvaTestConfig {
    constructor(filesPath, mocha, network, networkName, bail = false, timeout = 0, colors = false, verbose = false, isTest = false) {
        var _a;
        this.testingFiles = [];
        if (isTest) {
            if (filesPath && filesPath.length > 0) {
                if (fs_1.lstatSync(filesPath[0]).isDirectory()) {
                    filesPath.forEach(path => {
                        this.testingFiles.push(...fs_1.readdirSync(path_1.resolve(path)).map(f => path_1.join(path, f)));
                    });
                }
                else if (fs_1.lstatSync(filesPath[0]).isFile()) {
                    this.testingFiles = filesPath.map(f => path_1.resolve(f));
                }
            }
            else {
                if (isTest)
                    throw new Error('No files for test');
            }
        }
        this.mocha = mocha || {};
        this.bail = bail;
        this.verbose = verbose;
        this.timeout = timeout;
        this.networkList = require(network == null
            ? FindConfigFile_1.getConfigureModule(null)
            : path_1.resolve(network));
        this.halvaJs = require(network == null
            ? FindConfigFile_1.getConfigureModule(null)
            : path_1.resolve(network));
        this.types =
            ((_a = this.halvaJs.polkadotjs) === null || _a === void 0 ? void 0 : _a.types) == undefined
                ? null
                : this.halvaJs.polkadotjs.types;
        this.halvaJs = this.halvaJs.networks[networkName == null ? Object.keys(this.halvaJs.networks)[0] : networkName];
        this.mnemonic = this.halvaJs.mnemonic;
        this.networkName = networkName;
        this.colors = colors;
    }
}
exports.HalvaTestConfig = HalvaTestConfig;
