// @halva-suite/halva-cli/build/main/Deployer/Deployer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deployContract = exports.callContractRPC = exports.callContract = exports.instantiate = exports.UploadContract = void 0;
// tslint:disable: no-implicit-dependencies
// tslint:disable: no-submodule-imports
const api_1 = require("@polkadot/api");
const testing_1 = __importDefault(require("@polkadot/keyring/testing"));
const fs_1 = require("fs");
const consts_1 = require("./consts");
const utils_1 = require("./utils");
async function UploadContract(filePath, polkadot, account) {
    const tx = polkadot.tx.contracts.putCode(`0x${utils_1.GetByteArray(filePath)}`);
    const result = await utils_1.sendAndReturnFinalized(account, tx);
    const record = result.findRecord('contracts', 'CodeStored');
    if (!record) {
        console.error('\x1b[31m%s\x1b[0m', 'ERROR: No code stored after executing putCode()');
        console.warn('\x1b[31m%s\x1b[0m', 'Check block for more info: \n Block hash: ' + result.status.asFinalized);
        process.exit(126);
    }
    return record.event.data[0].toString();
}
exports.UploadContract = UploadContract;
exports.instantiate = async (api, signer, codeHash, inputData, endowment, gasRequired = consts_1.GAS_REQUIRED) => {
    const tx = api.tx.contracts.instantiate(endowment, gasRequired, codeHash, inputData);
    const result = await utils_1.sendAndReturnFinalized(signer, tx);
    const record = result.findRecord('contracts', 'Instantiated');
    if (!record) {
        console.error('\x1b[31m%s\x1b[0m', 'ERROR: No new instantiated contract');
        console.warn('\x1b[31m%s\x1b[0m', 'Check block for more info: \n Block hash: ' + result.status.asFinalized);
        process.exit(126);
    }
    // Return the Address of  the instantiated contract.
    return record.event.data[1];
};
exports.callContract = async (api, signer, contractAddress, inputData, gasRequired = consts_1.GAS_REQUIRED, endowment = 0) => {
    const tx = api.tx.contracts.call(contractAddress, endowment, gasRequired, inputData);
    const result = await utils_1.sendAndReturnFinalized(signer, tx);
    return result;
};
async function callContractRPC(api, signer, contractAddress, inputData, gasRequired = consts_1.GAS_REQUIRED, endowment = 0) {
    signer = signer;
    const rpc = await api.rpc.contracts.call({
        origin: signer.address,
        dest: contractAddress,
        value: endowment,
        gasLimit: gasRequired,
        inputData
    });
    if (rpc.isError) {
        throw new Error('RPC cal is error');
    }
    return rpc.asSuccess;
}
exports.callContractRPC = callContractRPC;
exports.deployContract = async (contract, abi, constructorIndex, args, config) => {
    const provider = new api_1.WsProvider(config.halvaJs.ws);
    const polkadot = await api_1.ApiPromise.create({ provider, types: config.types });
    const keyring = testing_1.default({ type: 'sr25519' });
    const alicePair = keyring.getPair(consts_1.ALICE);
    const hash = await UploadContract(contract, polkadot, alicePair);
    console.log('\x1b[33m%s\x1b[0m', 'WASM code hash ' + hash);
    const address = await exports.instantiate(polkadot, alicePair, hash, utils_1.GetAbiData(abi, constructorIndex, args), consts_1.CREATION_FEE);
    console.log('\x1b[33m%s\x1b[0m', `Contract address: ${address}`);
    return {
        address,
        abiJSON: fs_1.readFileSync(abi, 'utf-8'),
        abi: utils_1.getAbiObj(require(abi)),
        path: { contractPath: contract, AbiPath: abi }
    };
};
