// @halva-suite/halva-cli/build/main/Commands/migrate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Migrate = void 0;
const Migration_1 = require("../Migrator/Migration");
const TestRunner_1 = require("../TestRunner");
class Migrate {
    constructor() {
        this.command = 'migrate';
        this.describe = 'Run migrations';
    }
    builder(args) {
        return args
            .option('p', {
            alias: 'path',
            type: 'string',
            required: true,
            describe: 'Path to migrations folder'
        })
            .option('c', {
            alias: 'config',
            type: 'string',
            required: false,
            default: null,
            describe: 'Path to configure file'
        })
            .option('n', {
            alias: 'network',
            type: 'string',
            required: false,
            default: null,
            describe: 'Network name'
        });
    }
    async handler(args) {
        Migration_1.RunMigration(args.p, new TestRunner_1.HalvaTestConfig(null, null, args.c, args.n));
    }
}
exports.Migrate = Migrate;
