// @halva-suite/halva-cli/build/main/Commands/test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Test = void 0;
const TestRunner_1 = require("../TestRunner");
class Test {
    constructor() {
        this.command = 'test';
        this.describe = 'Run tests';
    }
    builder(args) {
        return args
            .option('p', {
            alias: 'path',
            type: 'string',
            default: './test/',
            array: true,
            required: false,
            describe: 'Path to test folder'
        })
            .option('c', {
            alias: 'config',
            type: 'string',
            required: false,
            default: null,
            describe: 'Path to configure file'
        })
            .option('n', {
            alias: 'network',
            type: 'string',
            required: false,
            default: null,
            describe: 'Network name'
        })
            .option('b', {
            alias: 'bail',
            type: 'boolean',
            required: false,
            default: null,
            describe: 'Enable bail'
        })
            .option('v', {
            alias: 'verbose',
            type: 'boolean',
            required: false,
            default: false,
            describe: 'Wite debug info'
        })
            .option('t', {
            alias: 'timeout',
            type: 'number',
            required: false,
            default: 0,
            describe: 'Timeout time for test'
        });
    }
    async handler(args) {
        TestRunner_1.HalvaRunTests(new TestRunner_1.HalvaTestConfig(args.p, null, args.c, args.n, args.d, args.t, null, args.v, true));
    }
}
exports.Test = Test;
