// @halva-suite/halva-cli/build/main/Migrator/Artifacts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractPrimitive = exports.artifacts = void 0;
const util_1 = require("@polkadot/util");
const fs_1 = require("fs");
const path_1 = require("path");
const Deployer_1 = require("../Deployer/Deployer");
const utils_1 = require("../Deployer/utils");
class artifacts {
    static require(contract) {
        const contractObj = JSON.parse(fs_1.readFileSync(process.cwd() + '/deployData.json', 'utf-8'));
        return this.CreateContractObject(contractObj, contract, globalThis.networkName);
    }
    static async ReturnTxResult(abiMethod, signer, contract, args) {
        const fn = contract.abi.messages[util_1.stringCamelCase(abiMethod)];
        let data;
        if (args == null) {
            data = fn();
        }
        else {
            data = fn(args);
        }
        return Deployer_1.callContract(halva.polkadot, signer, contract.address, data);
    }
    static async ReturnRPCResult(abiMethod, signer, contract, args) {
        const fn = contract.abi.messages[util_1.stringCamelCase(abiMethod)];
        let data;
        if (args == null) {
            data = fn();
        }
        else {
            data = fn(args);
        }
        return Deployer_1.callContractRPC(halva.polkadot, signer, contract.address, data);
    }
    static CreateContractObject(contract, contractName, networkName) {
        let deployed = false;
        const ContractData = contract.networks
            .find(x => x.networkName == networkName)
            .contracts.find(c => path_1.basename(c.path.contractPath, path_1.extname(c.path.contractPath)) ==
            contractName);
        if (ContractData) {
            deployed = true;
        }
        ContractData.abi = utils_1.getAbiObj(JSON.parse(ContractData.abiJSON));
        // @ts-ignore currently unused local ignore
        for (const [key, value] of Object.entries(ContractData.abi.abi.contract.messages)) {
            ContractPrimitive.prototype[value.name] = async function (signer, args) {
                const txResult = await artifacts.ReturnTxResult(value.name, signer, ContractData, args);
                const rpcResult = await artifacts.ReturnRPCResult(value.name, signer, ContractData, args);
                return { txResult, rpcResult };
            };
        }
        const ContractObject = new ContractPrimitive(deployed, contractName);
        return ContractObject;
    }
}
exports.artifacts = artifacts;
// tslint:disable-next-line: max-classes-per-file
class ContractPrimitive {
    constructor(deployed, contract) {
        this.deployed = deployed;
        this.contract = contract;
    }
    new() {
        return this.contract;
    }
    Deployed() {
        return this.deployed;
    }
}
exports.ContractPrimitive = ContractPrimitive;
