// @halva-suite/halva-cli/build/main/Migrator/Migration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RunMigration = exports.Migration = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const Deployer_1 = require("../Deployer/Deployer");
const DeployData_1 = require("./DeployData");
class Migration {
    constructor(config) {
        this.config = config;
    }
    async deploy(contract, args, constructorIndex = 0) {
        const data = await Deployer_1.deployContract(path_1.resolve(contract.contractPath), path_1.resolve(contract.AbiPath), constructorIndex, args, this.config);
        const network = this.parent.data.networks.find(f => f.networkName == this.config.networkName);
        if (network) {
            network.contracts.push(data);
        }
        else {
            this.parent.data.networks.push({
                networkName: this.config.networkName,
                networkAddress: this.config.halvaJs.ws,
                contracts: [data]
            });
        }
        return data;
    }
}
exports.Migration = Migration;
exports.RunMigration = async (migrationsPath, config) => {
    const migrations = fs_1.readdirSync(path_1.resolve(migrationsPath));
    const deployer = new Migration(config);
    deployer.data = new DeployData_1.DeployData();
    deployer.data.networks = [];
    deployer.parent = deployer;
    // tslint:disable-next-line: prefer-for-of
    for (let i = 0; i < migrations.length; i++) {
        const migration = require(path_1.join(path_1.resolve(migrationsPath), migrations[i]));
        await migration(deployer);
    }
    fs_1.writeFileSync(process.cwd() + '/deployData.json', JSON.stringify(deployer.data));
    process.exit(0);
};
