// @halva-suite/halva-cli/build/main/Init/InitMain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.templateRunner = exports.CreateProject = exports.InitMain = void 0;
const fs_1 = require("fs");
const hygen_1 = require("hygen");
const path_1 = require("path");
const Git = require('nodegit');
const Logger = require('hygen/lib/logger');
const defaultTemplates = path_1.join(__dirname, '../_templates');
const SubstrateRepo = 'https://github.com/substrate-developer-hub/substrate-node-template.git';
exports.InitMain = async (force, boxName, projectName) => {
    if (boxName) {
        // TODO
    }
    if (!projectName) {
        exports.templateRunner(['test', 'new', '--name', 'example'], '');
        exports.templateRunner(['init', 'new'], '');
    }
    if (projectName) {
        if (force) {
            await exports.CreateProject(projectName);
            exports.templateRunner(['init', 'tests'], `/${projectName}`);
            exports.templateRunner(['init', 'new'], `/${projectName}`);
        }
        else if (!force) {
            if (fs_1.existsSync(path_1.join(process.cwd(), projectName))) {
                console.log(projectName + ' directory already exists! Using --force');
                process.exit(0);
            }
            await exports.CreateProject(projectName);
            exports.templateRunner(['init', 'tests'], `/${projectName}`);
            exports.templateRunner(['init', 'new'], `/${projectName}`);
        }
    }
};
exports.CreateProject = async (projectName) => {
    try {
        fs_1.mkdirSync(path_1.join(process.cwd(), projectName));
        console.log('Clone the substrate repository, it can take a long time :)');
        await Git.Clone(SubstrateRepo, path_1.join(process.cwd(), projectName));
    }
    catch (err) {
        console.log(err);
        process.exit(0);
    }
};
exports.templateRunner = (argv, path) => {
    hygen_1.runner(argv, {
        templates: defaultTemplates,
        cwd: process.cwd() + path,
        logger: new Logger(console.log.bind(console)),
        createPrompter: () => require('enquirer'),
        exec: (action, body) => {
            const opts = body && body.length > 0 ? { input: body } : {};
            return require('execa').shell(action, opts);
        }
    });
};
