// @halva-suite/halva-cli/build/main/Substrate/Start.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Start = void 0;
const path_1 = require("path");
const child_process_1 = require("child_process");
const fs_1 = require("fs");
const halva_spec_builder_1 = require("@halva-suite/halva-spec-builder");
const KeyloggerMiddleware_1 = require("@halva-suite/halva-spec-builder/dist/middlewares/KeyloggerMiddleware");
const SPEC_FILE_NAME = 'halvaSpec.json';
const SUBSTRATE_TAG = 'v2.0.0';
exports.Start = async (pathSubstrate, config) => {
    if (await getCurrentBranch() !== SUBSTRATE_TAG) {
        await spawnProcess('git', ['checkout', SUBSTRATE_TAG]);
    }
    await spawnProcess('./scripts/init.sh');
    await spawnProcess('cargo', ['build']);
    try {
        await checkFile(pathSubstrate);
    }
    catch (err) {
        console.error(err.message);
        return;
    }
    await buildSpec(pathSubstrate);
    (await halva_spec_builder_1.HalvaSpecModifier.init(path_1.join(process.cwd(), SPEC_FILE_NAME), '1152921504606847000', 10)
        .setMnemonic(config.mnemonic)
        .apply(halva_spec_builder_1.balanceMiddleware)
        .apply(KeyloggerMiddleware_1.keyloggerMiddleware)
        .run()).output(path_1.join(process.cwd(), SPEC_FILE_NAME));
    await spawnProcess(path_1.resolve(pathSubstrate), [
        `--chain=./${SPEC_FILE_NAME}`,
        '--validator',
        '--alice',
        '--tmp',
    ], false);
};
async function checkFile(path) {
    return new Promise((resolve, reject) => {
        fs_1.stat(path, (err, stats) => {
            if (err) {
                return reject(err);
            }
            return resolve(stats);
        });
    });
}
async function getCurrentBranch() {
    return new Promise((resolve, reject) => {
        child_process_1.exec('git rev-parse --abbrev-ref HEAD', (err, stdout) => {
            if (err) {
                reject(err);
                return;
            }
            resolve(stdout.trim());
        });
    });
}
async function removeSpec(specPath) {
    return new Promise((resolve, reject) => {
        fs_1.unlink(specPath, async (err) => {
            if (err) {
                return reject(err);
            }
            resolve();
        });
    });
}
async function buildSpec(pathSubstrate) {
    const specPath = path_1.join(process.cwd(), SPEC_FILE_NAME);
    try {
        await checkFile(specPath);
        await removeSpec(specPath);
    }
    catch (e) { }
    let spec = fs_1.createWriteStream(specPath);
    await spawnProcess(pathSubstrate, ['build-spec', '--dev'], true, (chunk) => spec.write(chunk.toString()), undefined, undefined, spec.close);
}
function spawnProcess(command, args, silent = false, onStdoudData, onStderrData, onErr, onClose) {
    return new Promise((resolve, reject) => {
        const proc = child_process_1.spawn(command, args);
        proc.stdout.on('data', function (chunk) {
            if (!silent)
                console.log(chunk.toString());
            if (onStdoudData)
                onStdoudData(chunk);
        });
        proc.stderr.on('data', function (chunk) {
            if (!silent)
                console.log(chunk.toString());
            if (onStderrData)
                onStderrData(chunk);
        });
        proc.on('error', err => {
            if (onErr)
                onErr(err);
            reject(err);
        });
        proc.on('close', () => {
            if (onErr)
                onClose();
            resolve();
        });
    });
}
