// @halva-suite/halva-spec-builder/dist/middlewares/SessionMiddleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionMiddleware = void 0;
exports.SessionMiddleware = (context) => {
    var _a, _b, _c, _d, _e, _f;
    if (!((_c = (_b = (_a = context.jsonSchema.genesis) === null || _a === void 0 ? void 0 : _a.runtime) === null || _b === void 0 ? void 0 : _b.palletSession) === null || _c === void 0 ? void 0 : _c.keys))
        throw new Error('Invalid JSON');
    const pairsEd = context.ed25519Keys.getPairs();
    const pairsSr = context.sr25519Keys.getPairs();
    let key = [];
    for (let i = 0; i < pairsEd.length; i++) {
        key.push([
            pairsSr[i].address,
            pairsSr[i].address,
            {
                grandpa: pairsEd[i].address,
                babe: pairsSr[i].address,
                im_online: pairsSr[i].address,
                parachain_validator: pairsSr[i].address,
                authority_discovery: pairsSr[i].address,
            },
        ]);
    }
    for (let i = 0; i < key.length; i++) {
        (_f = (_e = (_d = context.jsonSchema.genesis) === null || _d === void 0 ? void 0 : _d.runtime) === null || _e === void 0 ? void 0 : _e.palletSession) === null || _f === void 0 ? void 0 : _f.keys.push(key[i]);
    }
    return context.jsonSchema;
};
