// @halva-suite/halva-cli/build/main/Networks/GetNetworkList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = exports.GetNetworks = void 0;
const fs_1 = require("fs");
exports.GetNetworks = (config) => {
    let networks = [];
    for (const [key, value] of Object.entries(config.networkList.networks)) {
        let x = value;
        x = x;
        networks.push(key);
    }
    let deployData;
    if (fs_1.existsSync(process.cwd() + '/deployData.json')) {
        deployData = JSON.parse(fs_1.readFileSync(process.cwd() + '/deployData.json').toString());
    }
    if (!deployData) {
        deployData = [];
        let nets_noDeploy;
        nets_noDeploy = [];
        networks.forEach(n => {
            nets_noDeploy.push({
                name: n,
                address: config.halvaJs.networks[n].ws,
                contracts: null
            });
        });
        return nets_noDeploy;
    }
    for (let i = 0; i < deployData.networks.length; i++) {
        if (!networks.includes(deployData.networks[i].networkName)) {
            networks.push(deployData.networks[i].networkName);
        }
    }
    if (networks.length == 0) {
        console.warn('Networks not found');
    }
    let networkList;
    networkList = [];
    networks.forEach(net => {
        let currentNetwork = deployData.networks.find(x => x.networkName == net);
        let currentContracts;
        currentContracts = [];
        currentNetwork.contracts.forEach(c => {
            currentContracts.push({
                name: c.path.contractPath,
                address: c.address.toString(),
                deployed: true
            });
        });
        networkList.push({
            name: net,
            address: currentNetwork.networkAddress,
            contracts: currentContracts
        });
    });
    return networkList;
};
exports.run = (config) => {
    const networkList = exports.GetNetworks(config);
    networkList.forEach(n => {
        console.log(`\nNetwork: ${n.name} : WS: ${n.address}`);
        if (n.contracts) {
            n.contracts.forEach(c => {
                console.log(`\nContract name: ${c.name} : Address: ${c.address} : Deployed ${c.deployed}`);
            });
        }
    });
};
